### PR TITLE
[SYCL][Fusion] Improve error messages on incompatible ND-ranges

### PR DESCRIPTION
**Pull request by victor-eds**
Show detailed error messages when users try to fuse kernels with incompatible ND-ranges, showing different errors for each different scenario. Also combine the validation and fusion logic to reduce the number of ND-ranges list traversals.

### CODE REVIEW

**Review by jopperm — APPROVED** [review]
Neat!

File: sycl-fusion/common/include/Kernel.h [hunk 1]
```diff
@@ -265,12 +265,18 @@ class NDRange {
   constexpr int getDimensions() const { return Dimensions; }
 
   bool hasSpecificLocalSize() const { return LocalSize != AllZeros; }
+  bool hasUniformWorkGroupSizes() const {
+    for (int I = 0; I < Dimensions; ++I) {
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply sycl-fusion/common/include/Kernel.h:270]: Done!

**Comment by victor-eds** [comment]
Failed `SYCL :: Graph/Explicit/basic_usm_mixed.cpp` test unrelated to this PR

**Comment by sommerlukas** [comment]
Merging this, failure seems unrelated, will follow up with the SYCL Graphs team.

**Pull request merged**